packages/peregrine/lib/store/actions/cart/asyncActions.js

```diff
@@ -336,6 +336,52 @@ export const getCartDetails = payload => {
     };
 };
 
+export const retrieveAndMergeCarts = payload =>
+    async function thunk(dispatch) {
+        const { mergeCarts, fetchCartId } = payload;
+        const sourceCartId = await retrieveCartId();
+        try {
+            await dispatch(removeCart());
+            await dispatch(
+                createCart({
+                    fetchCartId
+                })
+            );
+        } catch (error) {
+            dispatch(actions.mergeCarts.receive(error));
+            return thunk(...arguments);
+        }
+
+        try {
+            // errors can come from graphql and are not thrown
+            const destinationCartId = await retrieveCartId();
+            await dispatch(removeCart());
+            dispatch(
+                actions.mergeCarts.request(sourceCartId, destinationCartId)
+            );
+            const { data } = await mergeCarts({
+                variables: {
+                    destinationCartId,
+                    sourceCartId
+                }
+            });
+            const { mergeCarts: details } = data;
+
+            dispatch(actions.mergeCarts.receive({ details }));
+        } catch (error) {
+            dispatch(actions.mergeCarts.receive(error));
+
+            const shouldResetCart = !error.networkError && isInvalidCart(error);
+            if (shouldResetCart) {
+                // Delete the cached ID from local storage.
+                await dispatch(removeCart());
```

Comment:
We may have the same problem in #2378 here. Do you mind waiting for the fix? You could also just apply the change to the `isInvalidCart` function that I made in the PR.

Reply:
Removed usage of redux.